mypy/mixedtraverser.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional
+
+from mypy.nodes import (
+    MypyFile, Var, FuncItem, ClassDef, AssignmentStmt, OperatorAssignmentStmt, ForStmt, WithStmt,
+    CastExpr, TypeApplication, TypeAliasExpr, TypeVarExpr, TypedDictExpr, NamedTupleExpr,
+    EnumCallExpr, PromoteExpr, NewTypeExpr
+)
+from mypy.types import Type
+from mypy.traverser import TraverserVisitor
+from mypy.typetraverser import TypeTraverserVisitor
+
+
+class MixedTraverserVisitor(TraverserVisitor, TypeTraverserVisitor):
+    """Recursive traversal of both Node and Type objects."""
+
+    # Symbol nodes
+
+    def visit_var(self, var: Var) -> None:
+        self.visit_optional_type(var.type)
+
+    def visit_func(self, o: FuncItem) -> None:
+        super().visit_func(o)
+        self.visit_optional_type(o.type)
+
+    def visit_class_def(self, o: ClassDef) -> None:
+        super().visit_class_def(o)
+        info = o.info
+        if info:
+            for base in info.bases:
+                base.accept(self)
+            self.visit_optional_type(info.typeddict_type)
+
+    def visit_type_alias_expr(self, o: TypeAliasExpr) -> None:
+        super().visit_type_alias_expr(o)
+        o.type.accept(self)
+
+    def visit_type_var_expr(self, o: TypeVarExpr) -> None:
+        super().visit_type_var_expr(o)
+        o.upper_bound.accept(self)
+        for value in o.values:
+            value.accept(self)
+
+    def visit_typeddict_expr(self, o: TypedDictExpr) -> None:
+        super().visit_typeddict_expr(o)
+        self.visit_optional_type(o.info.typeddict_type)
+
+    def visit_namedtuple_expr(self, o: NamedTupleExpr) -> None:
+        super().visit_namedtuple_expr(o)
+        assert o.info.tuple_type
+        o.info.tuple_type.accept(self)
+
+    def visit__promote_expr(self, o: PromoteExpr) -> None:
+        super().visit__promote_expr(o)
+        o.type.accept(self)
+
+    def visit_newtype_expr(self, o: NewTypeExpr) -> None:
+        super().visit_newtype_expr(o)
+        self.visit_optional_type(o.old_type)
+
+    # Statements
+
+    def visit_assignment_stmt(self, o: AssignmentStmt) -> None:
+        super().visit_assignment_stmt(o)
+        self.visit_optional_type(o.type)
```

Comment:
Again, this type is not guaranteed to be the _same_ type as `TypeAlias.target` it technically may be a copy.

Reply:
If the file and line number is the same, it shouldn't be a problem. The main constraint is that we don't want to visit something that might be defined in another module (line number refer to the wrong file).

Reply:
The problem here might that the bad instance from `A = List[int, str]` will appear in two places, and only one will be fixed if we don't traverse both.

Reply:
I just noticed your other comment, if these are fixed before, then this should be fine.